### PR TITLE
Make Hibernate Reactive tests use @RunOnVertxContext and upgrade the Context validity checks

### PR DESCRIPTION
**Pull request by geoand**
[edited by @Sanne ]
This makes the internal integration tests of Hibernate Reactive use the `@RunOnVertxContext` annotation and also:

1. makes the `@RunOnVertxContext` also work for internal tests
2. introduces an attribute on @RunOnVertxContext to specify the kind of context
3. Changes/renames the `quarkus-junit5-vertx` - necessary to resolve cross-module dependencies
4. Upgrade the checks that Hibernate Reactive performs to validate it running on the right context: it will now fail by default if it's run on an un-blessed context.

This resolves all remaining concerns I had in regards to context usage for H.R.

N.B. This is a breaking change because the module name of `quarkus-junit5-vertx` was changed to `quarkus-test-vertx` (as it now needs to support `quarkus-junit5-internal`), the dependency on `quarkus-junit5` was removed and the API changed packages (from `io.quarkus.test.junit.vertx` to `io.quarkus.test.vertx`).

### CODE REVIEW

**Comment by cescoffier** [comment]
Because you are creating a root context and not a duplicated context :-)

**Comment by geoand** [comment]
I need the magic sauce to get to that duplicated context 😉

**Comment by geoand** [comment]
Nevermind, found it, so let me try

**Comment by geoand** [comment]
Success!

**Comment by geoand** [comment]
@Sanne all this PR needs now if the Hibernate tests to be ported (which is what the last commit started). Do you want to do that?

File: extensions/hibernate-reactive/deployment/src/test/java/io/quarkus/hibernate/reactive/SchemaUpdateTest.java [hunk 1]
```diff
@@ -28,12 +27,15 @@
...
     Mutiny.SessionFactory sessionFactory;
 
+    @RunOnVertxContext
     @Test
-    public void integerIdentifierWithStageAPI() {
-        final Uni<List<Object>> result = sessionFactory.withSession(s -> s
-                .createQuery("from Hero h where h.name = :name").setParameter("name", "Galadriel").getResultList());
-        final List<Object> list = result.await().indefinitely();
-        assertThat(list).isEmpty();
+    public void integerIdentifierWithStageAPI(UniAsserter asserter) {
```
Comment (cescoffier) [inline extensions/hibernate-reactive/deployment/src/test/java/io/quarkus/hibernate/reactive/SchemaUpdateTest.java:32]: Not sure I like that, but I understand why it goes in this direction. 

The problem is that the API is totally different from the UniAssertSubscriber from Mutiny.

File: extensions/hibernate-reactive/deployment/src/test/java/io/quarkus/hibernate/reactive/SchemaUpdateTest.java [hunk 2]
```diff
@@ -28,12 +27,15 @@
     @Inject
     Mutiny.SessionFactory sessionFactory;
 
+    @RunOnVertxContext
```
Comment (cescoffier) [inline extensions/hibernate-reactive/deployment/src/test/java/io/quarkus/hibernate/reactive/SchemaUpdateTest.java:30]: That's not enough, what kind of context is it running on?
Event loop context, worker context, duplicated context? 

The default should be a duplicated context, but we may need parameters to create the right type of context.

**Comment by Sanne** [comment]
Many thanks! Nice, there are some similarities with what I was toying with yesterday:
 - https://github.com/quarkusio/quarkus/compare/main...Sanne:CheckContext?expand=1

In particular I was preparing to introduce an attribute for the `@RunOnVertxContext` to control if the context shuld be duplicated, to make it more explicit. WDYT should we add it?

File: test-framework/vertx/src/main/java/io/quarkus/test/vertx/RunOnVertxContextTestMethodInvoker.java [hunk 3]
```diff
@@ -74,7 +75,7 @@ public Object invoke(Object actualTestInstance, Method actualTestMethod, List<Ob
         CompletableFuture<Object> cf = new CompletableFuture<>();
         RunTestMethodOnContextHandler handler = new RunTestMethodOnContextHandler(actualTestInstance, actualTestMethod,
                 actualTestMethodArgs, uniAsserter, cf);
-        vertx.getOrCreateContext().runOnContext(handler);
+        VertxContext.getOrCreateDuplicatedContext(vertx).runOnContext(handler);
```
Comment (cescoffier) [inline test-framework/vertx/src/main/java/io/quarkus/test/vertx/RunOnVertxContextTestMethodInvoker.java:78]: I think we need to give more flexibility. Yes, the default should be a duplicated context with its root being the event loop context, but we may need some other cases.

That being said, that should always run on a duplicated context.

**Comment by Sanne** [comment]
How important is the module rename commit?

If possible I'd refrain from that, so to not introduce breaking changes.

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: extensions/hibernate-reactive/deployment/src/test/java/io/quarkus/hibernate/reactive/SchemaUpdateTest.java (hunk 2 above)
Comment (Sanne) [reply extensions/hibernate-reactive/deployment/src/test/java/io/quarkus/hibernate/reactive/SchemaUpdateTest.java:30]: Right - likek I suggested I believe we should have some attributes on the annotation to make things explicit and controllable.

But I think "duplicated" could be a reasonable default.

https://github.com/quarkusio/quarkus/compare/main...Sanne:CheckContext?expand=1#diff-b4dcc8850c70f1ef3ed1c7c55fd3cd5b12db20e8b41f3cf8904e0d6ef3536333L77-R86

**Comment by geoand** [comment]
> If possible I'd refrain from that, so to not introduce breaking changes.

Not really possible to avoid unfortunately because of the dependency of the existing module on `quarkus-junit5`.

File: extensions/hibernate-reactive/deployment/src/test/java/io/quarkus/hibernate/reactive/SchemaUpdateTest.java (hunk 1 above)
Comment (geoand) [reply extensions/hibernate-reactive/deployment/src/test/java/io/quarkus/hibernate/reactive/SchemaUpdateTest.java:32]: Yes, it is... But I am not sure what we can do about it...

**Comment by geoand** [comment]
> In particular I was preparing to introduce an attribute for the @RunOnVertxContext to control if the context shuld be duplicated, to make it more explicit. WDYT should we add it?

Yeah, I think it makes sense. Feel free to make any changes you like in the PR :)

File: extensions/hibernate-reactive/deployment/src/test/java/io/quarkus/hibernate/reactive/SchemaUpdateTest.java (hunk 2 above)
Comment (Sanne) [reply extensions/hibernate-reactive/deployment/src/test/java/io/quarkus/hibernate/reactive/SchemaUpdateTest.java:30]: Cool, yes thanks a lot to implement the testing tool bits. I was trying yesterday but honestly got lost - not familiar with it at all so I appreciate you doing that part.

Comment (geoand) [reply extensions/hibernate-reactive/deployment/src/test/java/io/quarkus/hibernate/reactive/SchemaUpdateTest.java:30]: Yeah, it's too easy to get lost with all the necessary classloader hacks...

**Comment by Sanne** [comment]
I think this is done now - could @geoand and/or @cescoffier have another look?

**Comment by geoand** [comment]
LGTM

**Review by Sanne — APPROVED** [review]

**Comment by gsmet** [comment]
@Sanne does that mean that anyone using HR will have to annotate their tests?

**Comment by Sanne** [comment]
It should only affect users of the "internal" testing extension

**Comment by gsmet** [comment]
> It should only affect users of the "internal" testing extension

Ok, cool, good to know.

**Comment by Sanne** [comment]
And to be clear.. that's because all other uses already had the
restrictions. Perhaps people hadn't noticed, as resteasy reactive
implicitly sets up such context.

We only missed this one as it allows to sidestep other integration code,
it's now covered too.

**Review by cescoffier — APPROVED** [review]

**Comment by geoand** [comment]
I rebased onto main and force pushed in order to get CI going

**Comment by geoand** [comment]
Force pushed again to fix a bug I had introduced

**Comment by geoand** [comment]
Are we sure we want to backport this?

If we do, I understand the desire, just saying that this is a breaking change (albeit one that I likely affects very few people since `@RunOnVertxContext` isn't even mentioned in the docs)

**Comment by geoand** [comment]
Force pushed another fix

**Comment by Sanne** [comment]
@geoand for who is it actually a breaking change?

To be clear Hibernate Reactive always had the requirement to run on a vert.x context - it just so happens it was able to bypass this requirement in these specific integration tests because the use the **internal testing** tools which aren't meant for end users.

The only user-affecting change I see is that `@RunOnVertxContext` now executes (by default) on a duplicated context; but 

1. I don't expect that to have negative impact on tests - it probably was the original _intent_
2. The obscure annotation wasn't adverstised
3. We give the option to run on the global context instead

Let me know if you think I'm forgetting some aspect :)

**Comment by geoand** [comment]
It breaks anyone using `@RunOnVertxContext` because:

- The maven module that contains it changed name
- The dependencies of the maven module that contained it changed
- The package of the annotation changed

But as we both said, this is very unlikely to have been used by anybody as we never mentioned it anywhere.

**Comment by Sanne** [comment]
@geoand ah, sure. But you clarified that in the description already - I thought you meant there was more breaking stuff on top of it.

I think we're good in that case - perhaps I wouldn't have changed the module names as I said earlier, but you said you had no choice :) If we're doing what we can IMO we'll need to go ahead and ask for forgiveness.

**Comment by geoand** [comment]
Agreed

**Comment by edeandrea** [comment]
> But as we both said, this is very unlikely to have been used by anybody as we never mentioned it anywhere.

That may not be 100% true, as it has broken the super heroes sample app. It doesn't break people _just_ using the `@RunOnVertxContext`, but also breaks anyone using the `UniAsserter`, which would be pretty much anyone writing transactional-based tests with Hibernate Reactive.

**Comment by Sanne** [comment]
hi @edeandrea , could you point me to the test in the super heroes app that was affected?

It would help me to see a concrete example, as _think_ this only affects either internal testing API _or_ code which wasn't running in the right context to begin with. I might be wrong though :)

**Comment by edeandrea** [comment]
@Sanne https://github.com/quarkusio/quarkus-super-heroes/blob/main/rest-heroes/src/test/java/io/quarkus/sample/superheroes/hero/repository/HeroRepositoryTests.java (after I made changes).

I had to change the `artifactId` of the dependency in `pom.xml` as well and then update the java package of the `UniAsserter` class.

Here's the commit: https://github.com/quarkusio/quarkus-super-heroes/commit/5050d31bb04f8b6bf959a5b702963e108c238b4b

**Comment by cescoffier** [comment]
so, the problem is the artifactId change. @geoand do you believe we can to a relocation?

**Comment by edeandrea** [comment]
> so, the problem is the artifactId change. @geoand do you believe we can to a relocation?

Its not just the artifactId change - also the package relocation from `io.quarkus.test.junit.vert` to `io.quarkus.test.vertx`

**Comment by geoand** [comment]
Yeah, I didn't do the relocation thing because the package also had to change.

**Comment by Sanne** [comment]
I guess I don't fully follow why the package had to change? Could we have a deprecated copy of the annotation in the previous location?

**Comment by edeandrea** [comment]
`UniAsserter` isn't an annotation - but it probably could have been deprecated and either extended in a new package (or moved to a new package and extended/deprecated in the old)?

**Comment by geoand** [comment]
The old package didn't make sense after the repurposing of the API to work for both internal and user facing tests

**Comment by Sanne** [comment]
> The old package didn't make sense after the repurposing of the API to work for both internal and user facing tests

Ok but that's not a reason that it *had to* change ? I would suggest introducing a deprecated alternative for both the removed types.

**Comment by geoand** [comment]
There were too many changes that stacked up which is why I didn't see the reason for trying to keep compatibility

**Pull request merged**